src/Functions/reinterpretAs.cpp

```diff
@@ -176,18 +176,23 @@ class FunctionReinterpret : public IFunction
                     size_t size = offsets_from.size();
                     auto & vec_res = col_res->getData();
                     vec_res.resize_fill(size);
+                    size_t copy_size = 0;
+                    size_t offset_to = 0;
 
                     size_t offset = 0;
                     for (size_t i = 0; i < size; ++i)
                     {
+                        copy_size = std::min(static_cast<UInt64>(sizeof(ToFieldType)), offsets_from[i] - offset - 1);
+                        offset_to = sizeof(ToFieldType) > copy_size ? sizeof(ToFieldType) - copy_size : 0;
                         if constexpr (std::endian::native == std::endian::little)
                             memcpy(&vec_res[i],
                                 &data_from[offset],
-                                std::min(static_cast<UInt64>(sizeof(ToFieldType)), offsets_from[i] - offset - 1));
+                                copy_size);
                         else
-                            reverseMemcpy(&vec_res[i],
+                            reverseMemcpy(
+                                reinterpret_cast<char*>(&vec_res[i]) + offset_to,
                                 &data_from[offset],
-                                std::min(static_cast<UInt64>(sizeof(ToFieldType)), offsets_from[i] - offset - 1));
+                                copy_size);
```

Comment:
About l. 203: Below, I left a comment about similar changes potentially necessary for FixedString. The same would apply here - maybe you would like to check, thanks.